internal/terraform/eval_variable.go

```diff
@@ -90,27 +90,37 @@ func prepareFinalInputVariableValue(addr addrs.AbsInputVariableInstance, raw *In
 	val, err := convert.Convert(given, convertTy)
 	if err != nil {
 		log.Printf("[ERROR] prepareFinalInputVariableValue: %s has unsuitable type\n  got:  %s\n  want: %s", addr, given.Type(), convertTy)
+		var detail string
+		var subject *hcl.Range
 		if nonFileSource != "" {
-			diags = diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid value for input variable",
-				Detail: fmt.Sprintf(
-					"Unsuitable value for %s %s: %s.",
-					addr, nonFileSource, err,
-				),
-				Subject: cfg.DeclRange.Ptr(),
-			})
+			detail = fmt.Sprintf(
+				"Unsuitable value for %s %s: %s.",
+				addr, nonFileSource, err,
+			)
+			subject = cfg.DeclRange.Ptr()
 		} else {
-			diags = diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Invalid value for input variable",
-				Detail: fmt.Sprintf(
-					"The given value is not suitable for %s declared at %s: %s.",
-					addr, cfg.DeclRange.String(), err,
-				),
-				Subject: sourceRange.ToHCL().Ptr(),
-			})
+			detail = fmt.Sprintf(
+				"The given value is not suitable for %s declared at %s: %s.",
+				addr, cfg.DeclRange.String(), err,
+			)
+			subject = sourceRange.ToHCL().Ptr()
+		}
+
+		// In some workflows, the operator running terraform does not have access to the variables
+		// themselves. They are for example stored in encrypted files that will be used by the CI toolset
+		// and not by the operator directly. In such a case, the failing secret value should not be
+		// displayed to the operator
+		if cfg.Sensitive {
+			subject = nil
```

Comment:
I don't think there's any need to remove the subject here when it comes from a non-file source, and we probably also need to treat the logic for appending to the diagnostic detail message separately too. I think therefore we probably want an `if cfg.Sensitive { … }` block inside both branches of the `if nonFileSource != "" { … }` statement.

Reply:
that's how I coded it originally and changed it back to that.

But I read all your comments and they make sense to me. I'll try to take some time this week to update the PR